stock_reserve_rule/models/stock_reserve_rule.py

```diff
@@ -333,3 +398,68 @@ def _apply_strategy_full_bin(self, quants):
+        tolerance = self.tolerance_requested_value
+        limit = self.tolerance_requested_limit
+        computation = self.tolerance_requested_computation
+        if limit == "no_tolerance" or float_compare(tolerance, 0, rounding) == 0:
+            return float_compare(need, product_qty, rounding) == 0
+        elif limit == "upper_limit":
+            if computation == "percentage":
+                # need + rounding < product_qty <= need * (100 + tolerance) / 100
+                return (
+                    float_compare(need, product_qty, rounding) == -1
+                    and float_compare(
+                        product_qty, need * (100 + tolerance) / 100, rounding
+                    )
+                    <= 0
+                )
+            else:
+                # need + rounding < product_qty <= need + tolerance
+                return (
+                    float_compare(need, product_qty, rounding) == -1
+                    and float_compare(product_qty, need + tolerance, rounding) <= 0
+                )
+        elif limit == "lower_limit":
+            if computation == "percentage":
+                # need * (100 - tolerance) / 100 <= product_qty < need - rounding
+                return (
+                    float_compare(need * (100 - tolerance) / 100, product_qty, rounding)
+                    <= 0
+                    and float_compare(product_qty, need, rounding) == -1
+                )
+            # computation == "absolute"
+            else:
+                # need - tolerance <= product_qty < need - rounding
+                return (
+                    float_compare(need - tolerance, product_qty, rounding) <= 0
+                    and float_compare(product_qty, need, rounding) == -1
+                )
+
+    def _apply_strategy_full_lot(self, quants):
+        need = yield
+        # We take goods only if we empty the bin.
+        # The original ordering (fefo, fifo, ...) must be kept.
+        product = fields.first(quants).product_id
+        rounding = product.uom_id.rounding
+        if product.tracking == "lot":
+            for lot, lot_quants in quants.filtered(
+                lambda quant, product_id=product.id: quant.product_id.id == product_id
+            ).group_by_lot():
+                product_qty = sum(lot_quants.mapped("quantity"))
+                lot_quantity = sum(lot_quants.mapped("quantity")) - sum(
+                    lot_quants.mapped("reserved_quantity")
+                )
+                if (
+                    lot
+                    and self._compare_with_tolerance(need, product_qty, rounding)
+                    and lot_quantity > 0
+                ):
+                    need = (
+                        yield fields.first(lot_quants).location_id,
+                        lot_quantity,
+                        need,
+                        lot,
+                        None,
+                    )
```

Comment:
You have selected a lot and it can have multiple quants. I think you need to yield each quant one by one by looping on the quants (they should be already sorted by fifo/fefo) decreasing the need until need is 0 or all quants are consumed.
Can you add a test with this?
Example:
* Loc1, Lot A, 10 units
* Loc2, Lot A, 5 units
Need for 15 units. Lot A is selected

Reply:
@jbaudoux Thank you for your solution! I updated function implementation. Now the strategy looks better!